compute/disks/disks_test.go

```diff
@@ -263,66 +349,113 @@ func TestComputeDisksSnippets(t *testing.T) {
+
+		diskUrl := fmt.Sprintf("projects/%s/regions/%s/disks/%s", tc.ProjectID, region, instanceDiskName2)
+
+		if err := attachRegionalDisk(buf, tc.ProjectID, zone, instanceName, diskUrl); err != nil {
+			t.Fatalf("attachRegionalDisk got err: %v", err)
 		}
 		if got := buf.String(); !strings.Contains(got, want) {
-			t.Fatalf("setDiskAutoDelete got %q, want %q", got, want)
+			t.Fatalf("attachRegionalDisk got %q, want %q", got, want)
 		}
 
 		instance, err := getInstance(ctx, tc.ProjectID, zone, instanceName)
 		if err != nil {
 			t.Fatalf("getInstance got err: %v", err)
 		}
 
-		if instance.GetDisks()[0].GetAutoDelete() != true {
-			t.Errorf("instance got %t, want %t", instance.GetDisks()[0].GetAutoDelete(), true)
+		foundDisk := false
+		for _, disk := range instance.GetDisks() {
+			if strings.Contains(*disk.Source, instanceDiskName2) {
+				foundDisk = true
+			}
+		}
+		if !foundDisk {
+			t.Errorf("The disk %s is not attached to the instance!", instanceDiskName2)
+		}
+
+		// cannot clean up the disk just yet because it must be done after the VM is terminated
+	})
+
+	t.Run("Attach a read-only regional disk to VM", func(t *testing.T) {
+		if err := createRegionalDisk(buf, tc.ProjectID, region, replicaZones, instanceDiskName3, "regions/us-west3/diskTypes/pd-ssd", 20); err != nil {
+			t.Fatalf("createRegionalDisk got err: %v", err)
 		}
 
-		err = deleteInstance(ctx, tc.ProjectID, zone, instanceName)
+		buf.Reset()
+		want := "Disk attached"
+
+		diskUrl := fmt.Sprintf("projects/%s/regions/%s/disks/%s", tc.ProjectID, region, instanceDiskName3)
+
+		if err := attachRegionalDiskReadOnly(buf, tc.ProjectID, zone, instanceName, diskUrl); err != nil {
+			t.Fatalf("attachRegionalDisk got err: %v", err)
+		}
+		if got := buf.String(); !strings.Contains(got, want) {
+			t.Fatalf("attachRegionalDisk got %q, want %q", got, want)
+		}
+
+		instance, err := getInstance(ctx, tc.ProjectID, zone, instanceName)
 		if err != nil {
-			t.Errorf("deleteInstance got err: %v", err)
+			t.Fatalf("getInstance got err: %v", err)
 		}
+
+		foundDisk := false
+		for _, disk := range instance.GetDisks() {
+			if strings.Contains(*disk.Source, instanceDiskName3) {
+				foundDisk = true
+			}
+		}
+		if !foundDisk {
+			t.Errorf("The disk %s is not attached to the instance!", instanceDiskName3)
+		}
+
+		// cannot clean up the disk just yet because it must be done after the VM is terminated
 	})
 
 	// clean up
 	err = deleteDiskSnapshot(ctx, tc.ProjectID, snapshotName)
 	if err != nil {
 		t.Errorf("deleteDiskSnapshot got err: %v", err)
 	}
+
+	err = deleteInstance(ctx, tc.ProjectID, zone, instanceName)
+	if err != nil {
+		t.Errorf("deleteInstance got err: %v", err)
+	}
+
+	// disks attached to a VM instance must be deleted after deleting the instance
+
+	// disk 1 is set to auto-delete, no need to delete it explicitly
+
+	if err := deleteRegionalDisk(buf, tc.ProjectID, region, instanceDiskName2); err != nil {
+		t.Fatalf("deleteRegionalDisk got err: %v", err)
```

Comment:
These 2 `t.Fatalf()` should be `t.Errorf()`, so that problems with disk2 deletion do not also leave disk3 lying around.

FYI, you could also use [`t.Cleanup`](https://pkg.go.dev/testing#T.Cleanup) to ensure these resources are cleaned up, as our minimum go version is 1.15.

Reply:
Good catch, thanks!

I didn't want to use t.Cleanup because it doesn't report errors correctly, making it difficult to understand what went wrong when it fails:

https://github.com/golang/go/issues/51395
https://github.com/golang/go/issues/29207

Reply:
that's a good reason to avoid using t.Cleanup. thanks for the context links!